test/toolbox_image.sh

```diff
@@ -8,247 +8,65 @@
-    mount -t overlay overlay \
-          -o "lowerdir=${overlayfs_dir}/rom,upperdir=${overlayfs_dir}/up,workdir=${overlayfs_dir}/work" \
-          "${rootfs_dir}"
-
-    mkdir -p "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}"
-    tar xvfJ "${TEST_UPDATE_ROOTFS_FILE}" -C "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}" \
-        > /dev/null 2> /dev/null
-
-    update_mount="$(mktemp -d)"
-
-    touch "${rootfs_dir}/${ARM_EMU_BIN}"
-    mount --bind -o ro "${ARM_EMU_BIN}" "${rootfs_dir}/${ARM_EMU_BIN}"
-
-    mount --bind /proc "${rootfs_dir}/proc" 1> /dev/null
-    ln -s ../proc/self/mounts "${rootfs_dir}/etc/mtab"
-
-    if ! grep -qE "/dev.*devtmpfs" /proc/mounts; then
-        mount -t devtmpfs none "/dev"
-        is_dev_setup_mounted=true
-    fi
-
-    mount -t devtmpfs none "${rootfs_dir}/dev"
-
-    dd if=/dev/zero of="${rootfs_dir}/${TMP_TEST_IMAGE_FILE}" bs=1 count=0 seek=128M 2> /dev/null
+    toolbox_root_dir="$(mktemp -d -t "um-update-toolbox.XXXXXXXXXX")"
+    setup_chroot_env "${toolbox_image}" "${toolbox_root_dir}"
 
-    create_dummy_storage_device
-
-    sfdisk -d "${LOOP_STORAGE_DEVICE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}"
-
-    # Add the partition labels to the partition file, they will be used as label
-    sed -i "s|${LOOP_STORAGE_DEVICE}p1.*|&, name=boot|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
-    sed -i "s|${LOOP_STORAGE_DEVICE}p2.*|&, name=root|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
-    sed -i "s|${LOOP_STORAGE_DEVICE}p3.*|&, name=user|" "${rootfs_dir}${PARTITION_TABLE_FILE}"
+    dd if=/dev/zero of="${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}" bs=1 count=0 seek=128M
 }
 
 teardown()
 {
-    mounts="${rootfs_dir} ${overlayfs_dir}/rom ${overlayfs_dir}"
-
-    if [ ! -d "${rootfs_dir}" ]; then
-        return
-    fi
-
-    for partition_table_file in "${rootfs_dir}${PARTITION_TABLE_FILE}"*; do
-        unlink "${partition_table_file}"
-    done
-
-    if [ -d "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}" ]; then
-        rm -rf "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}"
-    fi
-
-    if [ -d "${update_mount}" ]; then
-        rm -r "${update_mount:?}"
-    fi
-
-    if [ -b "${LOOP_STORAGE_DEVICE}" ]; then
-        losetup -d "${LOOP_STORAGE_DEVICE}"
-    fi
-
-    if [ -f "${rootfs_dir}${STORAGE_DEVICE_IMG}" ]; then
-        unlink "${rootfs_dir}${STORAGE_DEVICE_IMG}"
-    fi
-
-    if [ -e "${rootfs_dir}/${ARM_EMU_BIN}" ]; then
-        if grep -q "$(realpath "${rootfs_dir}/${ARM_EMU_BIN}")" "/proc/mounts"; then
-            umount "${rootfs_dir}/${ARM_EMU_BIN}" || result=1
-        fi
-        if [ -f "${rootfs_dir}/${ARM_EMU_BIN}" ]; then
-            unlink "${rootfs_dir}/${ARM_EMU_BIN}" || result=1
-        fi
+    if [ -f "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}" ]; then
+        unlink "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}"
     fi
 
-    if grep -q "${rootfs_dir}/dev" /proc/mounts; then
-        umount "${rootfs_dir}/dev" || result=1
-    fi
+    teardown_chroot_env "${toolbox_root_dir}"
```

Comment:
I didn't spot this earlier, and didn't really think about this earlier, but this is a serious liability.

If, a test author, who is unaware of the consequences, puts this function in the wrong place, e.g. not at the top, you could in quite easily do:
```sh
rm -rf "${toolbox_root_dir}"
umount "${toolbox_root_dir}/proc"
```
by doing it in the wrong order, you first nuke the proc directory, and only after that try to umount it.

This can fail so horribly, in that it can bring down the server the CI runs on, as we are using privilidged mode, and mounting dev and proc from the host, rm -rf-ing those crashes the server.

Before, with the single teardown, this risk of course was also present, but it was more easy to oversee. Now, with several teardowns from each routine, this is a potential risk.